File: python/src/tokenize.py
class _EOF:
    def __repr__(self):
        return "#EOF#"

    def __len__(self):
        return 0

EOF = _EOF()
whitespace = [' ', '\n', '\r', '\t']
delimiters = ['(', ')', '[', ']', EOF] + whitespace


class Tokenizer:
    def __init__(self, source_string):
        self.original_source = source_string
        self.remaining = source_string
        self.current = ''
        self.last = ''

    def peek(self):
        current, remaining = _get_token(self.remaining)
        #print('peeked  : %s rest : %s' % (current, remaining))
        return current

    def chomp(self):
        self.last = self.current
        self.current, self.remaining = _get_token(self.remaining)
        #print('chomped : %s rest : %s' % (self.current, self.remaining))
        return self.current

    def to_pieces(self):
        def _get_piece(remaining, pieces):
            if remaining == EOF:
                return pieces
            else:
                self.chomp()
                return _get_piece(self.remaining, pieces + [self.current])
        return _get_piece(self.original_source, [])


def _get_token(remaining):
    def _get_next_character(remaining):
        if len(remaining) > 1:
            return remaining[0], remaining[1:]
        elif len(remaining) > 0:
            return remaining[0], EOF
        else:
            return EOF, EOF

    def _get_next_non_whitespace(remaining):
        current, remaining = _get_next_character(remaining)
        while current in whitespace:
            current, remaining = _get_next_character(remaining)
        return current, remaining

    current, remaining = _get_next_non_whitespace(remaining)
    token = current
    if not current in delimiters:
        while True:
            current, will_be_remaining = _get_next_character(remaining)
            if current in delimiters:
                break
            remaining = will_be_remaining
            token += current
    return token, remaining

File: haskell/src/Tokenize.hs
module Tokenize (
    Token(..),
    to_tokens
) where

import qualified Util

data Token
    = IntLiteral Int
    | BoolLiteral Bool
    | String_ String
    | Eof
    | LParen
    | RParen
    | LBracket
    | RBracket deriving (Show, Eq)

eof :: Char
eof = '\0'

whitespace :: [Char]
whitespace = [' ', '\n', '\r', '\t']

delimiters :: [Char]
delimiters = ['(', ')', '[', ']', eof] ++ whitespace

to_tokens :: String -> [Token]
to_tokens input_string =
    if (head input_string) == eof then
        []
    else
        let (token, remaining) = get_token input_string
        in
            token: to_tokens remaining

get_token :: String -> (Token, String)
get_token remaining =
    let (current, post_remaining) = get_next_non_whitespace remaining
    in
        if current `elem` delimiters then
            (convert_token [current], post_remaining)
        else
            build_next_token [current] post_remaining

convert_token :: String -> Token
convert_token current
    | current == "(" = LParen
    | current == ")" = RParen
    | current == "[" = LBracket
    | current == "]" = RBracket
    | current == [eof] = Eof
    | Util.is_int_literal current = IntLiteral $ read current
    | Util.is_bool_literal current   = BoolLiteral $ current == "true"
    | otherwise = String_ current


build_next_token :: String -> String -> (Token, String)
build_next_token input_token remaining =
    let (current, post_remaining) = get_next_character remaining
    in
        if current `elem` delimiters then
            (convert_token input_token, remaining)
        else
            build_next_token (input_token ++ [current]) post_remaining

get_next_non_whitespace :: String -> (Char, String)
get_next_non_whitespace remaining =
    let (current, post_remaining) = get_next_character remaining
    in
        if current `elem` whitespace then
            get_next_non_whitespace post_remaining
        else
            (current, post_remaining)

get_next_character :: String -> (Char, String)
get_next_character [] = (eof, [eof])
get_next_character [current] = (current, [eof])
get_next_character (current: rest) = (current, rest)
